File: windows/DoNotType.Core/FailureAdvice.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace DoNotType.Core;

/// <summary>
/// What went wrong, in a sentence, and what to do about it.
/// </summary>
/// <remarks>
/// <para>
/// Failures reached the interface as whatever was thrown: `HTTP 429: {"error":{"code":
/// "rate_limit_exceeded","message":"..."}}`, or a `HttpRequestException` naming a socket. That is a
/// log line. Somebody who has just spoken a sentence has two questions — is this my fault, and did
/// I lose what I said — and neither is answered by a status code.
/// </para>
/// <para>
/// The rules are the ones in `Sources/DoNotTypeCore/Reachability.swift`, deliberately identical:
/// the same failure on a laptop and a desktop should read the same way. docs/MANUAL-CHECKS.md sets
/// the standard — "the message should say the key is wrong, not 'The operation couldn't be
/// completed'".
/// </para>
/// </remarks>
public static class FailureAdvice
{
    /// <param name="Message">One line, shown in the overlay and stored on the history row.</param>
    /// <param name="IsQueued">Whether the dictation is safe and can be sent later.</param>
    /// <param name="IsRetryable">Whether retrying now is worth anything.</param>
    /// <param name="NeedsUserAction">
    /// Whether the user has to change something before this can ever work. False for a request
    /// this app got wrong: nothing in Settings fixes that, and sending somebody there when nothing
    /// they can change will help is worse than telling them it is not their fault.
    /// </param>
    public sealed record Guidance(
        string Message, bool IsQueued, bool IsRetryable, bool NeedsUserAction);

    public static Guidance Describe(Exception error, bool isOnline = true)
    {
        if (!isOnline)
        {
            return new Guidance(
                "Offline — saved, and it will send itself when you reconnect.",
                IsQueued: true, IsRetryable: true, NeedsUserAction: false);
        }

        if (error is ProviderException provider)
        {
            if (provider.Status > 0) return DescribeHttp(provider.Status, provider.Body);

            // No status: a parse failure, an empty output, or a message this app wrote itself.
            return new Guidance(
                provider.Message, IsQueued: true, IsRetryable: provider.IsTransient,
                NeedsUserAction: false);
        }

        if (error is AudioDecoder.DecodeException)
        {
            // Already written for a person, at the point of failure.
            return new Guidance(
                error.Message, IsQueued: false, IsRetryable: false, NeedsUserAction: true);
        }

        if (error is HttpRequestException or TaskCanceledException or IOException)
        {
            return new Guidance(
                "Network trouble — saved, and it will send itself when you reconnect.",
                IsQueued: true, IsRetryable: true, NeedsUserAction: false);
        }

        return new Guidance(
            error.Message, IsQueued: true, IsRetryable: true, NeedsUserAction: false);
    }

    private static Guidance DescribeHttp(int status, string body)
    {
        // xAI answers a bad key with 400 and a sentence about it, not with 401. Read by status
        // alone that lands in the default branch and becomes advice that cannot ever work, for a
        // request that will fail identically every time. Observed live.
        if (status == 400 && MentionsApiKey(body))
        {
            return new Guidance(
                "The API key was rejected. Check it in Settings.",
                IsQueued: false, IsRetryable: false, NeedsUserAction: true);
        }

        // What the provider itself said leads, when it said something readable, because it is
        // more specific than any status code can be — it knows what it refused. The advice follows.
        // The other way round buried "This model does not accept audio input." behind a sentence
        // about HTTP 400, and repeated ourselves when the provider had already said the same thing:
        // "The API key was rejected. Check it in Settings. Invalid API key provided."
        Guidance Say(string summary, string next, bool queued, bool retryable, bool needsAction)
        {
            var lead = Message(body) ?? summary;
            return new Guidance(
                next.Length == 0 ? lead : $"{lead} {next}", queued, retryable, needsAction);
        }

        return status switch
        {
            401 or 403 => Say(
                "The API key was rejected.", "Check it in Settings.",
                queued: false, retryable: false, needsAction: true),

            402 => Say(
                "Billing problem on the provider account — the key is valid but has no quota.", "",
                queued: false, retryable: false, needsAction: true),

            404 => Say(
                "That model is not available on this account.", "Pick another in Settings.",
                queued: false, retryable: false, needsAction: true),

            413 => Say(
                "The recording was too large for the provider.",
                "Long ones are normally split automatically, so this is worth reporting.",
                queued: false, retryable: false, needsAction: true),

            429 => Say(
                "Rate limited.", "Saved, and it will retry shortly.",
                queued: true, retryable: true, needsAction: false),

            408 => Say(
                "The provider took too long to answer.", "Saved, retry from History.",
                queued: true, retryable: true, needsAction: false),

            >= 500 and <= 599 => Say(
                "The provider is having trouble.", "Saved, retry from History.",
                queued: true, retryable: true, needsAction: false),

            // A 4xx is a request this app got wrong and will get wrong again in exactly the same
            // way, so it is kept but not offered as a retry.
            >= 400 and <= 499 => Say(
                $"The provider rejected the request (HTTP {status}).",
                "Retrying will not change it — this is likely a fault here, and worth reporting.",
                queued: true, retryable: false, needsAction: false),

            _ => Say(
                $"Request failed (HTTP {status}).", "Saved, retry from History.",
                queued: true, retryable: true, needsAction: false),
        };
    }

    /// <summary>
    /// Deliberately narrow. A 400 is normally a request this app got wrong, which is not the
    /// user's problem to fix — only one that names the key is reattributed to the key.
    /// </summary>
    private static bool MentionsApiKey(string body)
    {
        var lowered = body.ToLowerInvariant();
        return lowered.Contains("api key") || lowered.Contains("api_key")
            || lowered.Contains("apikey");
    }

    /// <summary>The human-readable part of an error body, if there is one.</summary>
    /// <remarks>
    /// Every OpenAI-compatible provider answers with `{"error": {"message": "..."}}`, and that
    /// sentence is routinely the most useful thing available. Parsed rather than printed raw, so a
    /// body that is not a sentence — a trace ID, an HTML error page, a wall of JSON — is dropped
    /// instead of pasted into the corner of somebody's screen. A user reading an overlay is not
    /// debugging.
    /// </remarks>
    internal static string? Message(string body)
    {
        var trimmed = body.Trim();
        if (trimmed.Length == 0) return null;

        try
        {
            if (JsonNode.Parse(trimmed) is { } parsed) return Tidy(MessageIn(parsed));
        }
        catch (JsonException)
        {
            // Not JSON. A short plain-text body is usually a gateway saying something useful; a
            // long one, or one starting a tag, is an error page.
            if (trimmed.Length <= 200 && !trimmed.StartsWith('<')) return Tidy(trimmed);
        }
        return null;
    }

    /// <summary>The shapes providers actually use.</summary>
    private static string? MessageIn(JsonNode? node)
    {
        if (node is JsonValue value && value.TryGetValue<string>(out var text))
        {
            return string.IsNullOrEmpty(text) ? null : text;
        }
        if (node is not JsonObject dictionary) return null;

        foreach (var key in new[] { "message", "error_description", "detail" })
        {
            if (dictionary[key] is JsonValue candidate
                && candidate.TryGetValue<string>(out var found)
                && !string.IsNullOrEmpty(found))
            {
                return found;
            }
        }
        foreach (var key in new[] { "error", "err", "failure" })
        {
            if (dictionary[key] is { } nested && MessageIn(nested) is { } deeper) return deeper;
        }
        return null;
    }

    /// <summary>One line, ending in a full stop, short enough to sit in a pill on a screen.</summary>
    private static string? Tidy(string? text)
    {
        if (text is null) return null;

        var flattened = string.Join(
            " ",
            text.Split('\n', '\r', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries));
        if (flattened.Length == 0) return null;

        var capped = flattened.Length <= 140 ? flattened : flattened[..137].TrimEnd() + "…";

        // It leads the message now, so it starts a sentence, and gateways answer in lower case.
        // Only a plain word is capitalised: the first token is often an identifier the provider is
        // quoting back — `models/gemini-9 is not found` — and "Models/gemini-9" is a different name.
        var firstWord = capped.Split(' ')[0];
        var opened = firstWord.All(char.IsLetter)
            ? char.ToUpperInvariant(capped[0]) + capped[1..]
            : capped;
        return opened.EndsWith('.') || opened.EndsWith('…') ? opened : opened + ".";
    }
}

File: android/app/src/main/kotlin/app/donottype/core/FailureAdvice.kt
package app.donottype.core

import app.donottype.audio.AudioDecoder
import java.io.IOException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import org.json.JSONObject

/**
 * What went wrong, in a sentence, and what to do about it.
 *
 * Failures reached the keyboard as whatever was thrown: `HTTP 429: {"error":{"code":
 * "rate_limit_exceeded","message":"…"}}`, or the class name of an exception. That is a log line.
 * Somebody who has just spoken a sentence has two questions — is this my fault, and did I lose what
 * I said — and neither is answered by a status code, least of all on a phone where the message has
 * one line to fit in.
 *
 * The rules are the ones in `Sources/DoNotTypeCore/Reachability.swift` and
 * `windows/DoNotType.Core/FailureAdvice.cs`, deliberately identical: the same failure on a phone
 * and a laptop should read the same way. `docs/MANUAL-CHECKS.md` sets the standard — "the message
 * should say the key is wrong, not 'The operation couldn't be completed'".
 */
object FailureAdvice {

    /**
     * @param message one line, shown on the keyboard and stored on the history row.
     * @param isQueued whether the dictation is safe and can be sent later.
     * @param isRetryable whether retrying now is worth anything.
     * @param needsUserAction whether the user has to change something before this can ever work.
     *   False for a request this app got wrong: nothing in Settings fixes that, and sending
     *   somebody there when nothing they can change will help is worse than telling them it is not
     *   their fault.
     */
    data class Guidance(
        val message: String,
        val isQueued: Boolean,
        val isRetryable: Boolean,
        val needsUserAction: Boolean,
    )

    /**
     * Errors worth retrying, as opposed to ones that will fail identically forever.
     *
     * Retrying a bad API key just burns the user's time; a timeout or a 503 is exactly what retry
     * exists for.
     */
    fun isTransient(error: Throwable): Boolean = when (error) {
        is SocketTimeoutException, is UnknownHostException -> true
        is ProviderException -> when {
            // The status, not a substring of the message. Asking whether the text contained
            // "HTTP 4" searched the provider's own body too, so a 500 whose body quoted an
            // upstream "HTTP 404" was classified as permanent and never retried.
            error.status > 0 -> error.status == 408 || error.status == 429 || error.status >= 500
            error.message.orEmpty().contains("billed 0 audio tokens") -> false
            else -> true
        }
        is IOException -> true
        else -> false
    }

    fun describe(error: Throwable, isOnline: Boolean = true): Guidance {
        if (!isOnline) {
            return Guidance(
                "Offline — saved, and it will send itself when you reconnect.",
                isQueued = true, isRetryable = true, needsUserAction = false,
            )
        }

        return when {
            error is ProviderException && error.status > 0 ->
                describeHttp(error.status, error.body)

            // No status: a parse failure, an empty output, or a message this app wrote itself.
            error is ProviderException -> Guidance(
                error.message ?: "The request failed.",
                isQueued = true, isRetryable = true, needsUserAction = false,
            )

            error is AudioDecoder.DecodeException -> Guidance(
                // Already written for a person, at the point of failure.
                error.message ?: "That recording could not be read.",
                isQueued = false, isRetryable = false, needsUserAction = true,
            )

            error is UnknownHostException || error is SocketTimeoutException || error is IOException ->
                Guidance(
                    "Network trouble — saved, and it will send itself when you reconnect.",
                    isQueued = true, isRetryable = true, needsUserAction = false,
                )

            else -> Guidance(
                error.message ?: error::class.simpleName.orEmpty(),
                isQueued = true, isRetryable = true, needsUserAction = false,
            )
        }
    }

    private fun describeHttp(status: Int, body: String): Guidance {
        // xAI answers a bad key with 400 and a sentence about it, not with 401. Read by status
        // alone that lands in the default branch and becomes advice that cannot ever work, for a
        // request that will fail identically every time. Observed live.
        if (status == 400 && mentionsApiKey(body)) {
            return Guidance(
                "The API key was rejected. Check it in Settings.",
                isQueued = false, isRetryable = false, needsUserAction = true,
            )
        }

        // What the provider itself said leads, when it said something readable, because it is
        // more specific than any status code can be — it knows what it refused. The advice follows.
        // The other way round buried "This model does not accept audio input." behind a sentence
        // about HTTP 400, and repeated ourselves when the provider had already said the same thing:
        // "The API key was rejected. Check it in Settings. Invalid API key provided."
        fun say(
            summary: String,
            next: String,
            queued: Boolean,
            retryable: Boolean,
            needsAction: Boolean,
        ): Guidance {
            val lead = message(body) ?: summary
            return Guidance(
                if (next.isEmpty()) lead else "$lead $next",
                isQueued = queued,
                isRetryable = retryable,
                needsUserAction = needsAction,
            )
        }

        return when (status) {
            401, 403 -> say(
                "The API key was rejected.", "Check it in Settings.",
                queued = false, retryable = false, needsAction = true,
            )

            402 -> say(
                "Billing problem on the provider account — the key is valid but has no quota.", "",
                queued = false, retryable = false, needsAction = true,
            )

            404 -> say(
                "That model is not available on this account.", "Pick another in Settings.",
                queued = false, retryable = false, needsAction = true,
            )

            413 -> say(
                "The recording was too large for the provider.",
                "Long ones are normally split automatically, so this is worth reporting.",
                queued = false, retryable = false, needsAction = true,
            )

            429 -> say(
                "Rate limited.", "Saved, and it will retry shortly.",
                queued = true, retryable = true, needsAction = false,
            )

            408 -> say(
                "The provider took too long to answer.", "Saved, retry from History.",
                queued = true, retryable = true, needsAction = false,
            )

            in 500..599 -> say(
                "The provider is having trouble.", "Saved, retry from History.",
                queued = true, retryable = true, needsAction = false,
            )

            // A 4xx is a request this app got wrong and will get wrong again in exactly the same
            // way, so it is kept but not offered as a retry.
            in 400..499 -> say(
                "The provider rejected the request (HTTP $status).",
                "Retrying will not change it — this is likely a fault here, and worth reporting.",
                queued = true, retryable = false, needsAction = false,
            )

            else -> say(
                "Request failed (HTTP $status).", "Saved, retry from History.",
                queued = true, retryable = true, needsAction = false,
            )
        }
    }

    /**
     * Deliberately narrow. A 400 is normally a request this app got wrong, which is not the user's
     * problem to fix — only one that names the key is reattributed to the key.
     */
    private fun mentionsApiKey(body: String): Boolean {
        val lowered = body.lowercase()
        return lowered.contains("api key") || lowered.contains("api_key") ||
            lowered.contains("apikey")
    }

    /**
     * The human-readable part of an error body, if there is one.
     *
     * Every OpenAI-compatible provider answers with `{"error": {"message": "…"}}`, and that
     * sentence is routinely the most useful thing available. Parsed rather than printed raw, so a
     * body that is not a sentence — a trace ID, an HTML error page, a wall of JSON — is dropped
     * instead of pasted onto somebody's keyboard. A user reading an error is not debugging.
     */
    internal fun message(body: String): String? {
        val trimmed = body.trim()
        if (trimmed.isEmpty()) return null

        if (trimmed.startsWith("{")) {
            return runCatching { messageIn(JSONObject(trimmed)) }.getOrNull()?.let(::tidy)
        }

        // Not JSON. A short plain-text body is usually a gateway saying something useful; a long
        // one, or one starting a tag, is an error page.
        if (trimmed.length > 200 || trimmed.startsWith("<")) return null
        return tidy(trimmed)
    }

    /** The shapes providers actually use. */
    private fun messageIn(json: JSONObject): String? {
        for (key in listOf("message", "error_description", "detail")) {
            val text = json.optString(key)
            if (text.isNotEmpty()) return text
        }
        for (key in listOf("error", "err", "failure")) {
            json.optJSONObject(key)?.let { nested -> messageIn(nested)?.let { return it } }
            val text = json.optString(key)
            if (text.isNotEmpty()) return text
        }
        return null
    }

    /** One line, ending in a full stop, short enough for a phone. */
    private fun tidy(text: String): String? {
        val flattened = text.split('\n', '\r')
            .map { it.trim() }
            .filter { it.isNotEmpty() }
            .joinToString(" ")
        if (flattened.isEmpty()) return null

        val capped = if (flattened.length <= 140) flattened else flattened.take(137).trimEnd() + "…"

        // It leads the message now, so it starts a sentence, and gateways answer in lower case.
        // Only a plain word is capitalised: the first token is often an identifier the provider is
        // quoting back — `models/gemini-9 is not found` — and "Models/gemini-9" is a different name.
        val opened = if (capped.substringBefore(' ').all { it.isLetter() }) {
            capped.replaceFirstChar { it.uppercase() }
        } else {
            capped
        }
        return if (opened.endsWith(".") || opened.endsWith("…")) opened else "$opened."
    }
}
